File: backend-python/app/models/spot_work.py
from sqlalchemy import BigInteger, Column, DateTime, ForeignKey, Index, String, Text
from sqlalchemy.dialects.postgresql import JSONB
from sqlalchemy.orm import relationship
from sqlalchemy.sql import func

from app.database import Base
from app.models.mixins import SoftDeleteMixin, SerializationMixin


class SpotWork(Base, SoftDeleteMixin, SerializationMixin):
    __tablename__ = "spot_work"

    _relation_overrides = {
        'client_name': ('project', 'client_name'),
    }
    _detail_relation_overrides = {
        'project_name': ('project', 'project_name'),
        'client_name': ('project', 'client_name'),
    }
    _relation_extras = {
        'address': ('project', 'address'),
        'client_contact_position': ('project', 'client_contact_position'),
    }

    id = Column(BigInteger, primary_key=True, autoincrement=True, comment="主键ID")
    work_id = Column(String(50), unique=True, nullable=False, comment="用工单编号")
    plan_id = Column(String(50), ForeignKey('maintenance_plan.plan_id', ondelete='CASCADE'), nullable=True, index=True, comment="关联维保计划编号")
    project_id = Column(String(50), ForeignKey('project_info.project_id', ondelete='CASCADE'), nullable=False, comment="项目编号")
    project_name = Column(String(200), nullable=False, comment="项目名称")
    plan_start_date = Column(DateTime, nullable=False, comment="计划开始日期")
    plan_end_date = Column(DateTime, nullable=False, comment="计划结束日期")
    client_name = Column(String(100), comment="客户单位")
    client_contact = Column(String(100), comment="客户联系人")
    client_contact_info = Column(String(50), comment="客户联系电话")
    maintenance_personnel = Column(String(100), comment="运维人员")
    work_content = Column(Text, comment="工作内容")
    photos = Column(JSONB, default=list, comment="现场图片JSON数组")
    signature = Column(Text, comment="班组签字图片")
    status = Column(String(20), nullable=False, default="执行中", comment="状态")
    remarks = Column(String(500), comment="备注")
    actual_completion_date = Column(DateTime, comment="实际完成时间")
    reject_reason = Column(String(500), comment="退回原因")
    created_at = Column(DateTime, server_default=func.now(), nullable=False, comment="创建时间")
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now(), nullable=False, comment="更新时间")

    project = relationship("ProjectInfo", back_populates="spot_works")
    maintenance_plan = relationship("MaintenancePlan", back_populates="spot_works")

    _list_exclude_fields = {'photos', 'signature', 'reject_reason'}

    def to_list_dict(self) -> dict:
        return self.to_dict(exclude=self._list_exclude_fields)

    __table_args__ = (
        Index('idx_spot_work_id', 'work_id'),
        Index('idx_spot_project_id', 'project_id'),
        Index('idx_spot_project_name', 'project_name'),
        Index('idx_spot_client_name', 'client_name'),
        Index('idx_spot_status', 'status'),
        Index('idx_spot_plan_start_date', 'plan_start_date'),
        Index('idx_spot_status_created', 'status', 'created_at'),
        Index('idx_spot_status_updated', 'status', 'updated_at'),
        {'comment': '零星用工单表'}
    )

File: packages/shared/src/types/models/spotWork.ts
/**
 * 零星用工数据模型
 */

/**
 * 零星用工工单
 */
export interface SpotWork {
  id: number
  work_id: string
  plan_id?: string
  plan_type?: string
  project_id: string
  project_name: string
  plan_start_date: string
  plan_end_date: string
  client_name?: string
  client_contact?: string
  client_contact_info?: string
  address?: string
  client_contact_position?: string
  maintenance_personnel?: string
  work_content?: string
  photos?: string[] | string
  signature?: string
  status: string
  remarks?: string
  actual_completion_date?: string
  worker_count?: number
  work_days?: number
  workers?: SpotWorkWorker[]
  created_at: string
  updated_at: string
}

/**
 * 零星用工工人
 */
export interface SpotWorkWorker {
  id: number
  spot_work_id?: number
  project_id?: string
  project_name?: string
  start_date?: string
  end_date?: string
  name: string
  gender?: string
  birth_date?: string
  address?: string
  id_card_number?: string
  issuing_authority?: string
  valid_period?: string
  id_card_front?: string
  id_card_back?: string
  created_at?: string
  updated_at?: string
}

/**
 * 创建零星用工请求
 */
export interface SpotWorkCreate {
  work_id?: string
  plan_id?: string
  project_id: string
  project_name: string
  plan_start_date: string
  plan_end_date: string
  client_name?: string
  client_contact?: string
  client_contact_info?: string
  maintenance_personnel?: string
  work_content?: string
  photos?: string | string[]
  signature?: string
  status?: string
  remarks?: string
  worker_count?: number
}

/**
 * 更新零星用工请求
 */
export interface SpotWorkUpdate {
  work_id: string
  plan_id?: string
  project_id: string
  project_name: string
  plan_start_date: string
  plan_end_date: string
  client_name?: string
  client_contact?: string
  client_contact_info?: string
  maintenance_personnel?: string
  work_content?: string
  photos?: string[] | string
  signature?: string
  status: string
  remarks?: string
  actual_completion_date?: string
  worker_count?: number
  work_days?: number
}

/**
 * 零星用工查询参数
 */
export interface SpotWorkQueryParams {
  page?: number
  size?: number
  project_name?: string
  work_id?: string
  status?: string
}

/**
 * 快速填报请求
 */
export interface QuickFillRequest {
  project_id: string
  project_name: string
  plan_start_date: string
  plan_end_date: string
  work_content?: string
  client_contact?: string
  client_contact_info?: string
  photos?: string
  signature?: string
  worker_count?: number
  remark?: string
}

/**
 * 工人信息保存请求
 */
export interface WorkersSaveRequest {
  spot_work_id?: number
  project_id?: string
  project_name?: string
  start_date?: string
  end_date?: string
  workers: Partial<SpotWorkWorker>[]
}
